contracts/src/v0.8/shared/token/ERC677/BurnMintERC677.sol

```diff
@@ -1,217 +1,33 @@
-    if (recipient == address(this)) revert();
-    _;
-  }
-
-  // ================================================================
-  // |                      Burning & minting                       |
-  // ================================================================
-
-  /// @inheritdoc ERC20Burnable
-  /// @dev Uses OZ ERC20 _burn to disallow burning from address(0).
-  /// @dev Decreases the total supply.
-  function burn(uint256 amount) public override(IBurnMintERC20, ERC20Burnable) onlyBurner {
-    super.burn(amount);
-  }
-
-  /// @inheritdoc IBurnMintERC20
-  /// @dev Alias for BurnFrom for compatibility with the older naming convention.
-  /// @dev Uses burnFrom for all validation & logic.
-  function burn(address account, uint256 amount) public virtual override {
-    burnFrom(account, amount);
-  }
-
-  /// @inheritdoc ERC20Burnable
-  /// @dev Uses OZ ERC20 _burn to disallow burning from address(0).
-  /// @dev Decreases the total supply.
-  function burnFrom(address account, uint256 amount) public override(IBurnMintERC20, ERC20Burnable) onlyBurner {
-    super.burnFrom(account, amount);
-  }
-
-  /// @inheritdoc IBurnMintERC20
-  /// @dev Uses OZ ERC20 _mint to disallow minting to address(0).
-  /// @dev Disallows minting to address(this)
-  /// @dev Increases the total supply.
-  function mint(address account, uint256 amount) external override onlyMinter validAddress(account) {
-    if (i_maxSupply != 0 && totalSupply() + amount > i_maxSupply) revert MaxSupplyExceeded(totalSupply() + amount);
-
-    _mint(account, amount);
-  }
-
-  // ================================================================
-  // |                            Roles                             |
-  // ================================================================
-
-  /// @notice grants both mint and burn roles to `burnAndMinter`.
-  /// @dev calls public functions so this function does not require
-  /// access controls. This is handled in the inner functions.
-  function grantMintAndBurnRoles(address burnAndMinter) external {
-    grantMintRole(burnAndMinter);
-    grantBurnRole(burnAndMinter);
-  }
-
-  /// @notice Grants mint role to the given address.
-  /// @dev only the owner can call this function.
-  function grantMintRole(address minter) public onlyOwner {
-    if (s_minters.add(minter)) {
-      emit MintAccessGranted(minter);
-    }
-  }
-
-  /// @notice Grants burn role to the given address.
-  /// @dev only the owner can call this function.
-  function grantBurnRole(address burner) public onlyOwner {
-    if (s_burners.add(burner)) {
-      emit BurnAccessGranted(burner);
-    }
-  }
-
-  /// @notice Revokes mint role for the given address.
-  /// @dev only the owner can call this function.
-  function revokeMintRole(address minter) public onlyOwner {
-    if (s_minters.remove(minter)) {
-      emit MintAccessRevoked(minter);
+contract BurnMintERC677 is BurnMintERC20, IERC677 {
```

Comment:
Can we inherit `ERC677` here so that we can avoid re-defining `transferAndCall()` below?

Reply:
No, because that means double importing ERC20 constructor which the compiler will not allow